sys/phydat/phydat_json.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_phydat
+ * @{
+ *
+ * @file
+ * @brief       Convert phydat_t structs to human readable JSON strings
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "fmt.h"
+#include "assert.h"
+#include "phydat.h"
+
+#define STATIC_LEN      (14U)
+
+/**
+ * @note    @p buf must be at least 5 bytes of size
+ */
+static size_t _bool_to_str(int16_t val, char *buf)
+{
+    if (val) {
+        memcpy(buf, "true", 4);
+        return 4;
+    }
+    else {
+        memcpy(buf, "false", 5);
+        return 5;
+    }
+}
+
+size_t phydat_to_json(const phydat_t *data, size_t dim, char *buf)
+{
+    assert((dim > 0) && (dim < PHYDAT_DIM));
+
+    size_t pos = 0;
+
+    if (buf == NULL) {
+        pos = STATIC_LEN;
+        if (dim > 1) {
+            pos += (2 + (dim - 1));       /* array parens + separating commas */
+        }
+        for (size_t i = 0; i < dim; i++) {
+            if (data->unit != UNIT_BOOL) {
+                pos += fmt_s16_dfp(NULL, data->val[i], (int)data->scale);
+            }
+            else {
+                pos += (data->val[i]) ? 4 : 5;  /* true: 4, false: 5 */
+            }
+        }
+        pos += strlen(phydat_unit_to_str_verbose(data->unit));
+    }
+    else {
+        memcpy(buf, "{\"d\":", 5);
+        pos += 5;
+        /* write data */
+        if (dim > 1) {
+            buf[pos++] = '[';
+        }
+        for (size_t i = 0; i < dim; i++) {
+            if (data->unit != UNIT_BOOL) {
```

Comment:
Nit-pick: reverse logic. It is easier for the brain to parse to handle the `data->unit == UNIT_BOOL` first (the connection to the logical expression to the called function's name is much closer)

Reply:
But as the first case is the more probable case, the present ordering has a higher chance to result in a binary that saves an instruction for this case ;-)

Reply:
I think all our toolchains are smart enough by now to reorder it as we want... But fine... not dying on that hill ;-).

Reply:
not quite, as the toolchain can not judge the probability here... The key impact is the CPU: on anything desktop-ish it does indeed not matter as the branch prediction in the CPU will make all paths equally fast by caching them. Also some RIOT platforms (aka m3, m4) do have some form of branch prediction. But the less advanced platforms (avr8, m0) do not support such a thing and hence manual ordering of conditions has an impact, though arguably not a significant one...